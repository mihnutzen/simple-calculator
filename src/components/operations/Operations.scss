@import 'src/styles/variables';

.Operations {
  display: grid;
  grid-template-columns: 1;
  grid-template-rows: repeat($operations-grid-rows, $height);

  grid-gap: 1px;
  background-color: #666;

  > button {
    background: $equal-blue;
    line-height: $height;

    &.selected {
      border: 2px solid black;
    }
  }
}