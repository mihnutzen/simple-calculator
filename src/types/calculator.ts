export enum OperationsList {
  Divide = 'divide',
  Multiply = 'multiply',
  Subtract = 'subtract',
  Addition = 'addition',
  Equal = 'equal'
}

export enum OperationsSigns {
  Divide = '/',
  Multiply = '*',
  Subtract = '-',
  Addition = '+',
  Equal = '='
}

export enum ActionsList {
  Clear = 'clear',
  Negate = 'negate',
  Percent = 'percent'
}

export enum ActionsSigns {
  Clear = 'AC',
  Negate = 'Â±',
  Percent = '%'
}

export type NumOpt = number | '.'