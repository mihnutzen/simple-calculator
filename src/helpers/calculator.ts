import { OperationsList, NumOpt } from '../types/calculator';

import * as self from './calculator';


export const getIsSelectedOperation = (op: OperationsList, selection: OperationsList | undefined) => {
  return op === selection && selection !== OperationsList.Equal;
}

export const getCurrentValue = (current: string, number: NumOpt) => {
  if (current !== '0') {
    return current + number;
  }

  if (current === '0' && number !== '.') {
    return number + '';
  }

  return current + number;
}

export const getDisplayValue = (prev: string, current: string) => {
  if (prev !== '0' && current === '0') {
    return prev;
  }

  return current;
}

export const getNegated = (val: string) => {
  if (val[0] === '0' && val.length < 3) {
    return val[0];
  }

  return val[0] === '-' ? val.substring(1) : '-' + val;
}

export const getAddition = (val1: number, val2: number) => val1 + val2;
export const getDivision = (val1: number, val2: number) => val2 / val1;
export const getMultiplication = (val1: number, val2: number) => val1 * val2;
export const getSubtraction = (val1: number, val2: number) => val2 - val1;

export const getCalculation = (val1: string, val2: string, op: OperationsList | undefined) => {
  const operationsMap = {
    [OperationsList.Addition]: self.getAddition,
    [OperationsList.Divide]: self.getDivision,
    [OperationsList.Multiply]: self.getMultiplication,
    [OperationsList.Subtract]: self.getSubtraction,
    [OperationsList.Equal]: () => {},
  }

  if (op) {
    return operationsMap[op](parseFloat(val1), parseFloat(val2)) + '';
  }

  return val2;
}