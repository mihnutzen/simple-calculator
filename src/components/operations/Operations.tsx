import React from 'react';
import classNames from 'classnames';

import { getIsSelectedOperation } from '../../helpers/calculator';
import { OperationsList, OperationsSigns } from '../../types/calculator';

import './Operations.scss';

const DIVIDE = {
  key: OperationsList.Divide,
  sign: OperationsSigns.Divide,
  helper: 'Divide (or press /)'
};

const MULTIPLY = {
  key: OperationsList.Multiply,
  sign: OperationsSigns.Multiply,
  helper: 'Multiply (or press *)'
};

const SUBTRACT = {
  key: OperationsList.Subtract,
  sign: OperationsSigns.Subtract,
  helper: 'Subtract (or press -)'
};

const ADDITION = {
  key: OperationsList.Addition,
  sign: OperationsSigns.Addition,
  helper: 'Add (or press +)'
};

const EQUAL = {
  key: OperationsList.Equal,
  sign: OperationsSigns.Equal,
  helper: 'Equal (or press Return)'
};

const OPERATIONS = [DIVIDE, MULTIPLY, SUBTRACT, ADDITION, EQUAL];

interface OperationsProps {
  onOperation: (op: OperationsList) => void;
  selectedOperation?: OperationsList;
}

const Operations = ({ onOperation, selectedOperation }: OperationsProps) => (
  <div className="Operations">
      {OPERATIONS.map(({ key, sign, helper }) => (
        <button
          className={classNames({ selected: getIsSelectedOperation(key, selectedOperation) })}
          onClick={() => onOperation(key)} key={key}
          title={helper}
        >
          {sign}
        </button>
      ))}
  </div>
);

export default Operations;