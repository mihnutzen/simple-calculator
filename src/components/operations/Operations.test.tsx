import React from 'react';
import { render, fireEvent } from '@testing-library/react';

import { OperationsList } from '../../types/calculator';

import Operations from './Operations';

describe('The Operations component', () => {
  it('should render correctly', () => {
    const operationMock = jest.fn();

    const { asFragment } = render(<Operations onOperation={operationMock} selectedOperation={OperationsList.Addition} />);
    const firstRender = asFragment()

    expect(firstRender).toMatchSnapshot();
  });

  it('should trigger onOperation with payload', () => {
    const operationMock = jest.fn();

    const { getByText } = render(<Operations onOperation={operationMock} />);

    fireEvent.click(getByText(/-/))

    expect(operationMock).toHaveBeenCalledWith('subtract');
  });
});
