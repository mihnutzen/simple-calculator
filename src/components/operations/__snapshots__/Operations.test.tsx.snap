// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Operations component should render correctly 1`] = `
<DocumentFragment>
  <div
    class="Operations"
  >
    <button
      class=""
      title="Divide (or press /)"
    >
      /
    </button>
    <button
      class=""
      title="Multiply (or press *)"
    >
      *
    </button>
    <button
      class=""
      title="Subtract (or press -)"
    >
      -
    </button>
    <button
      class="selected"
      title="Add (or press +)"
    >
      +
    </button>
    <button
      class=""
      title="Equal (or press Return)"
    >
      =
    </button>
  </div>
</DocumentFragment>
`;
